cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
   return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
 }
 
-/**
- * @brief Functor for detecting character escape sequences in URL-encoded strings.
- */
-struct url_decode_escape_detector {
-  device_span<char const> const d_chars{};
-
-  /**
-   * @brief Detects if an escape sequence is at the specified character position.
-   *
-   * Returns true for a character index corresponding to the start of an escape
-   * sequence, i.e.: '%' followed by two hexadecimal digits. This does not check
-   * against string boundaries and therefore can produce false-positives.
-   *
-   * @param char_idx Character position to examine
-   * @return true An escape sequence was detected at the character position
-   * @return false No escape sequence at the character position
-   */
-  __device__ bool operator()(size_type char_idx) const
-  {
-    return (char_idx + 2 < d_chars.size()) && d_chars[char_idx] == '%' &&
-           is_hex_digit(d_chars[char_idx + 1]) && is_hex_digit(d_chars[char_idx + 2]);
-  }
-};
+__forceinline__ __device__ bool is_escape_char(char const* const ptr)
+{
+  return (ptr[0] == '%' && is_hex_digit(ptr[1]) && is_hex_digit(ptr[2]));
+}
 
 /**
- * @brief Functor for detecting escape sequence positions that cross a string boundary.
+ * @brief Count the number of characters of each string after URL decoding.
+ *
+ * @tparam threadblock_size Number of threads in a threadblock. This template argument must be a
+ * multiple of `cudf::detail::warp_size`.
+ * @tparam block_size Number of characters which will be loaded into the shared memory at a time.
+ *
+ * @param[in] in_chars Character buffer for the input string column.
+ * @param[in] in_offsets Offset value of each string associated with `in_chars`.
+ * @param[out] out_counts Number of characters in each decode URL.
+ * @param[in] num_strings Number of strings to count the number of decoded characters.
  */
-struct url_decode_esc_position_filter {
-  device_span<int32_t const> const d_offsets{};
-
-  /**
-   * @brief Detects if an escape sequence crosses a string boundary
-   *
-   * Returns true for an escape sequence that straddles a boundary between two strings.
-   *
-   * @param esc_pos_idx Character position corresponding to the start of an escape sequence
-   * @return true The escape sequence crosses a string boundary
-   * @return false The escape sequence does not cross a string boundary
-   */
-  __device__ bool operator()(size_type esc_pos_idx) const
-  {
-    // find the end offset of the current string
-    size_type const* offset_ptr =
-      thrust::upper_bound(thrust::seq, d_offsets.begin(), d_offsets.end(), esc_pos_idx);
-    return esc_pos_idx + 2 >= *offset_ptr;
+template <int threadblock_size, int block_size>
```

Comment:
Can these both be `unsigned int` instead of `int`? I notice a lot of `int` variables in this kernel that I think could be unsigned unless they're being passed to CUDA APIs somewhere that I missed, for instance `warps_per_threadblock` and the five ints on lines 222-226, and most indexers for `for` loops. Could you please do a quick pass and convert the variables that don't need a sign bit? Many of them can also probably be declared with `auto`.

Reply:
Could you elaborate on why unsigned is preferred in this case? I think since both CUDA indices (e.g. threadIdx.x) and cuDF `size_type` and `offset_type` are signed type, it's probably a good idea to stay consistent to those.

Reply:
Good point about cuDF constants, I agree. [threadIdx is unsigned](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#threadidx), and in practice all of these represent unsigned constants (for example you can never really have a negative `warp_lane`, `nwarps`, or `*_warp_id`), but you're right that we may as well keep this signed if we're doing any arithmetic with signed types like the cudf constants.